lib/watir/locators/button/selector_builder/xpath.rb

```diff
@@ -3,68 +3,57 @@ module Locators
     class Button
       class SelectorBuilder
         class XPath < Element::SelectorBuilder::XPath
-          def build(selector)
-            return super if selector.key?(:adjacent)
-
-            selector[:tag_name] = 'button'
-
-            type = selector.delete :type
-            return super(selector) if type.eql?(false)
-
-            # both :value and :text selectors will locate elements by value attribute or text content
-            selector[:text] = selector.delete(:value) if selector.key?(:value)
-
-            wd_locator = super(selector)
-
-            start_string = default_start
-
-            button_string = "local-name()='button'"
-            common_string = wd_locator[:xpath].gsub(start_string, '').gsub("[#{button_string}]", '')
-
-            input_string = "(local-name()='input' and #{input_types(type)})"
-
-            tag_string = if type.nil?
-                           "[#{button_string} or #{input_string}]"
-                         else
-                           "[#{input_string}]"
-                         end
+          private
 
-            xpath = "#{start_string}#{tag_string}#{common_string}"
+          def tag_string
+            return super if @adjacent
 
-            {xpath: xpath}
-          end
+            @selector.delete(:tag_name)
 
-          protected
+            type = @selector.delete :type
 
-          def use_index?
-            false
+            if type
+              "[(local-name()='input' and #{input_types(type)})]"
+            elsif type.nil?
+              "[local-name()='button' or (local-name()='input' and #{input_types(type)})]"
+            else
+              "[local-name()='button']"
+            end
           end
 
           # This is special because text locator for buttons match text or value
-          def add_text
-            return '' unless @selector.key?(:text)
+          def text_string
```

Comment:
We need to consider the tag name when deciding to check text or value:
* For button elements, `:text` only matches the element text
* For input elements, `:text` only matches the value attribute

Given the following button, which has different text and value attribute:

~~~~~~~~
<button id="btn_dif_value" value="btn_dif_value">btn_dif_text</button>
~~~~~~~~

With these changes, `:text` will now match the text and value:

~~~~~~~~
browser.buttons(text: 'btn_dif_value').map(&:id)
#=> ["btn_dif_value"]
browser.buttons(text: 'btn_dif_text').map(&:id)
#=> ["btn_dif_value"]
~~~~~~~~

Where as 6.14 would only match the text:

~~~~~~~~
browser.buttons(text: 'btn_dif_value').map(&:id)
#=> []
browser.buttons(text: 'btn_dif_text').map(&:id)
#=> ["btn_dif_value"]
~~~~~~~~

Reply:
Ah, if it is an actual button we need to check the text only, and if it is an input we need to check the value only. Got it.

Reply:
@jkotests ok, so what is the right answer here:

```html
<button name="button_2" type="submit" value="button_2">Button 2</button>
```

```ruby
expect(browser.button(value: 'Button 2')).to exist
expect(browser.button(value: /Button 2/)).to exist
```

I kind of feel like this should only match `button_2` and not `Button 2`

I guess I should keep it backward compatible for now, though. :-/

Reply:
Oh. I don't think I ever noticed that. Yeah, I agree, it should be only matching the value attribute. More stuff for the list of things to deprecate.

Reply:
I made it backward compatible. I'll add it to the list of things to deprecate. :)

Reply:
It's like this logic is complicated or something...